### PR TITLE
Refactor workflow

### PR DESCRIPTION
**Pull request by artivis**
Refactor the overall workflow to split the main build/test/pull jobs.
This also unify the regular & lxc based testing job in a single file. 
Plus it supports the case where multiple snaps must be tested together (build & pub are parallelised).

### CODE REVIEW

File: .github/workflows/monitor-upstream.yaml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+          echo "upstream_version=$(./get-upstream-version.sh)" >> "$GITHUB_OUTPUT"
+
+      - name: Compare versions
+        id: compare
+        if: "${{ inputs.script-compare-version == '' }}"
+        env:
+          SNAP_VERSION: ${{ steps.parse.outputs.version }}
+          UPSTREAM_VERSION: ${{ steps.get-upstream-version.outputs.upstream-version }}
+        run: |
+          curl https://raw.githubusercontent.com/Ariel-Rodriguez/sh-semversion-2/main/semver2.sh -o semver2.sh
```
Comment (MirkoFerrati) [inline .github/workflows/monitor-upstream.yaml:87]: "curl https://raw.githubusercontent.com/Ariel-Rodriguez/sh-semversion-2/main/semver2.sh -o semver2.sh " @artivis this is a supply-chain risk. 
if the repo is taken over they could run scripts in our CI, which may also imply in our self-hosted runners.  Could you pinpoint a specific tag so we don't get new commits in the script we download?

Comment (artivis) [reply .github/workflows/monitor-upstream.yaml]: I removed this file as well as the other two that don't belong to this pr.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+          path: '${{ matrix.source-subdir }}'
+          snapcraft-channel: '${{ inputs.snapcraft-channel }}'
+          snapcraft-args: '${{ inputs.snapcraft-args }}'
+
+      - name: Set unique artifact name
+        id: artifact-unique-name
+        run: |
+          echo "artifact-name=workflow-build-snap-$(basename -s .snap ${{ steps.build-snap.outputs.snap }})" >> "$GITHUB_OUTPUT"
+
+        #  echo "artifact-name=$(basename -s .snap ${{ steps.build-snap.outputs.snap }})-$(echo ${{ inputs.git-ref }} | sed "s|\/|\-|g")" >> "$GITHUB_OUTPUT"
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/build.yaml:90]: To remove.

File: .github/workflows/publish.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      matrix: ${{ steps.get-snap-file.outputs.matrix }}
+    steps:
+
+      - name: Check Snapstore login
+        env:
+          SNAPCRAFT_STORE_CREDENTIALS: ${{ secrets.snapstore-login }}
+        run: |
+          # Check if the key is empty
+          if [ -z "${SNAPCRAFT_STORE_CREDENTIALS}" ]; then
+            echo "Snapstore login is empty !"
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/publish.yaml:29]:
```suggestion
            echo "Snapstore login is empty!"
```

File: .github/workflows/publish.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            exit 1
+          fi
+
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/publish.yaml:66]: This is already present in the prepare publish.

File: .github/workflows/snap-lxc.yaml [hunk 5]
```diff
@@ -1,139 +0,0 @@
-name: build-install-test-snap-lxc
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap-lxc.yaml:1]: Deleting this file will break current workflows. Should we make a workflow with this name calling the main one?

File: .github/workflows/snap.yaml [hunk 6]
```diff
@@ -3,173 +3,118 @@ name: snap
...
+        default: '${{ github.ref }}'
+        description: The branch to checkout.
         required: false
         type: string
-        default: "latest/stable"
-      snap-name:
-        required: true
+      lxd-image:
+        default: ''
+        description: The LXC image to run this action in.
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/snap.yaml:18]:
```suggestion
        description: The LXD image to run this action in.
```

File: .github/workflows/test.yaml [hunk 7]
```diff
@@ -0,0 +1,145 @@
+name: test
+
+on:
+  workflow_call:
+    inputs:
+      lxd-image:
+        default: ''
+        description: The LXC image to run this action in.
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/test.yaml:8]:
```suggestion
        description: The LXD image to run this action in.
```

File: .github/workflows/test.yaml [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      - name: Find out the arch
+        id: get-arch
+        run: echo "arch=$(dpkg --print-architecture)" >> "$GITHUB_OUTPUT"
+
+      - name: Download snap artifact(s)
+        uses: actions/download-artifact@v4
+        with:
+          path: .
+          pattern: 'workflow-build-snap-*${{ steps.get-arch.outputs.arch }}*'
+          # merge-multiple: true
```
Comment (Guillaumebeuzeboc) [inline .github/workflows/test.yaml:57]: To remove?

File: README.md [hunk 9]
```diff
@@ -1,33 +1,153 @@
...
+- [build.yaml](.github/workflows/build.yaml) - the workflow to build the snap.
+- [test.yaml](.github/workflows/test.yaml) - the workflow to test the snap.
+- [publish.yaml](.github/workflows/publish.yaml) - the workflow to publish the snap.
+
+### The snap workflow
+
+The [snap](.github/workflows/snap.yaml) workflow is the main reusable workflow.
+It calls, in one go, the build,
+test and release workflows as an integrated and coherent sequence.
+As such, it exposes all configurations options of each sub-workflow as detailed in the next section.
```
Comment (Guillaumebeuzeboc) [inline README.md:50]:
```suggestion
As such, it exposes all configuration options of each sub-workflow as detailed in the next section.
```

File: README.md [hunk 10]
```diff
@@ -1,33 +1,153 @@
...
+
+For further configurations, see each sub-workflow details below.
+
+#### Options
+
+| Option | Default Value | Description | Required |
+|---|---|---|---|
+| `cleanup` | false | Whether to cleanup the artifacts. | false |
+| `git-ref` | ${{ github.ref }} | The branch to checkout. | false |
+| `lxd-image` | ' ' | The LXC image to run this action in. | false |
```
Comment (Guillaumebeuzeboc) [inline README.md:67]:
```suggestion
| `lxd-image` | ' ' | The LXD image to run this action in. | false |
```

File: README.md [hunk 11]
```diff
@@ -1,33 +1,153 @@
...
+| `snapcraft-channel` | latest/stable | The channel from which to install Snapcraft. | false |
+| `snapcraft-enable-experimental-extensions` | false | Whether to enable Snapcraft experimental extensions or not. | false |
+| `snapcraft-source-subdir` | ' . ' | The path where to execute snapcraft. | false |
+
+### The build workflow
+
+The [build](.github/workflows/build.yaml) workflow,
+as it names suggests,
+builds the snap.
+It does so using the [canonical/action-build](https://github.com/canonical/action-build) and the snap project it find at the repository root.
```
Comment (Guillaumebeuzeboc) [inline README.md:82]:
```suggestion
It does so using the [canonical/action-build](https://github.com/canonical/action-build) and the snap project it finds at the repository root.
```

File: README.md [hunk 12]
```diff
@@ -1,33 +1,153 @@
...
+As such, its input is all the artefacts created during the run,
+out of which is uses all the `*.snap` file it finds.
+This allows for testing multiple snaps at once,
+especially convenient for intricate multi-snaps deployments.
+
+Similarly to the `build` workflow, callers can define the runner it is executed on.
+Note that it will only install the snap(s) of matching architecture.
+
+By default the workflow solely make sure the snap(s) is installable and calls `snapcraft info` on it.
+However the caller may provide an additional custom test as a bash script through the `snap-test-script` option.
```
Comment (Guillaumebeuzeboc) [inline README.md:114]:
```suggestion
However, the caller may provide an additional custom test as a bash script through the `snap-test-script` option.
```

File: string-to-json/action.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
+name: String to JSON
+description: Convert a string to a JSON list
+# author: ubuntu-robotics-community
+# branding:
+#   icon: trending-up
+#   color: orange
```
Comment (Guillaumebeuzeboc) [inline string-to-json/action.yaml:6]: to remove?

File: .github/workflows/build.yaml (hunk 2 above)
Comment (artivis) [reply .github/workflows/build.yaml]: Done.

File: string-to-json/action.yaml (hunk 13 above)
Comment (artivis) [reply string-to-json/action.yaml:6]: I haven't check what they correspond to yet. I'd leave it as a reminder.

File: .github/workflows/snap.yaml [hunk 14]
```diff
@@ -3,173 +3,118 @@ name: snap
...
+        default: '${{ github.ref }}'
+        description: The branch to checkout.
         required: false
         type: string
-        default: "latest/stable"
-      snap-name:
-        required: true
+      lxd-image:
+        default: ''
+        description: The LXD image to run this action in.
```
Comment (giusebar) [inline .github/workflows/snap.yaml:18]: this is only to run test in, maybe we should make it clearer here.

File: .github/workflows/build.yaml [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+jobs:
+  prepare-build-runners:
+    runs-on: ubuntu-latest
+    outputs:
+      runs-on: ${{ steps.string-to-json.outputs.json }}
+    steps:
+      - name: String to JSON list
+        id: string-to-json
+        uses: canonical/robotics-actions-workflows/string-to-json@fix/refactor-matrix
```
Comment (giusebar) [inline .github/workflows/build.yaml:45]: change branch before merging

File: .github/workflows/build.yaml [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+          string: ${{ inputs.runs-on }}
+
+  prepare-build-matrix:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.string-to-json.outputs.json }}
+    steps:
+      - name: String to JSON list
+        id: string-to-json
+        uses: canonical/robotics-actions-workflows/string-to-json@fix/refactor-matrix
```
Comment (giusebar) [inline .github/workflows/build.yaml:56]: change branch before merging

File: .github/workflows/test.yaml [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+jobs:
+  prepare-test-runners:
+    runs-on: ubuntu-latest
+    outputs:
+      runs-on: ${{ steps.string-to-json.outputs.json }}
+    steps:
+      - name: String to JSON list
+        id: string-to-json
+        uses: canonical/robotics-actions-workflows/string-to-json@fix/refactor-matrix
```
Comment (giusebar) [inline .github/workflows/test.yaml:35]: change branch before merging

File: .github/workflows/snap.yaml [hunk 18]
```diff
@@ -3,173 +3,118 @@ name: snap
...
         run: |
-          if jq -e . >/dev/null 2>&1 <<<"$input"; then
-              echo "list-of-params=$input" >> $GITHUB_OUTPUT
+          if [ "${{ secrets.snapstore-login }}" != '' ]; then
+            echo "defined=true" >> $GITHUB_OUTPUT;
           else
-              input_as_json="[\"$input\"]"
-              echo "ubuntu-image input is not a json, formatting it into a json list: $input_as_json."
-              echo "list-of-params=$input_as_json" >> $GITHUB_OUTPUT
+            echo "defined=false" >> $GITHUB_OUTPUT;
```
Comment (giusebar) [inline .github/workflows/snap.yaml:99]: should we fail and exit here?

File: .github/workflows/snap-lxc.yaml (hunk 5 above)
Comment (giusebar) [reply .github/workflows/snap-lxc.yaml:1]: it should be only these two running on core18: 
- https://github.com/snapcraft-docs/ros-talker-listener/blob/master/.github/workflows/snap.yaml
- https://github.com/snapcraft-docs/ros2-talker-listener/blob/master/.github/workflows/snap.yaml#L16

**Review by giusebar — CHANGES_REQUESTED** [review]

File: .github/workflows/publish.yaml [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        uses: actions/download-artifact@v4
+        with:
+          path: .
+          pattern: 'workflow-build-snap-*'
+
+      - name: Retrieve snap file
+        id: get-snap-file
+        run: |
+          files=$(find . -name '*.snap' -printf "\"%h/%f\",")
+          echo "Files: ${files}"
```
Comment (giusebar) [inline .github/workflows/publish.yaml:43]: should this be removed or wnat to leave it there for debugging?

File: .github/workflows/snap-lxc.yaml (hunk 5 above)
Comment (Guillaumebeuzeboc) [reply .github/workflows/snap-lxc.yaml:1]: Ok then it makes sens to break and fix

File: README.md [hunk 20]
```diff
@@ -1,33 +1,153 @@
...
+| `snap-track` | latest | Snap Store channel track use for publication. | false |
+| `snapcraft-args` | ' ' | The arguments to pass to snapcraft (pack). | false |
+| `snapcraft-channel` | latest/stable | The channel from which to install Snapcraft. | false |
+| `snapcraft-enable-experimental-extensions` | false | Whether to enable Snapcraft experimental extensions or not. | false |
+| `snapcraft-source-subdir` | ' . ' | The path where to execute snapcraft. | false |
+
+### The build workflow
+
+The [build](.github/workflows/build.yaml) workflow,
+as it names suggests,
```
Comment (giusebar) [inline README.md:80]:
```suggestion
as its names suggests,
```

File: README.md [hunk 21]
```diff
@@ -1,33 +1,153 @@
...
+- `snapcraft-args`
+- `snapcraft-channel`
+- `snapcraft-enable-experimental-extensions`
+- `snapcraft-source-subdir`
+
+### The test workflow
+
+The [test](.github/workflows/test.yaml) workflow aims at testing the snap(s) built by the `build` workflow.
+As such, its input is all the artefacts created during the run,
+out of which is uses all the `*.snap` file it finds.
```
Comment (giusebar) [inline README.md:106]:
```suggestion
out of which it uses all the `*.snap` file it finds.
```

File: README.md [hunk 22]
```diff
@@ -1,33 +1,153 @@
...
+The `test` uses the following subset of options from the `snap` workflow:
+
+- `lxd-image`
+- `runs-on`
+- `snap-install-args`
+- `snap-test-script`
+
+### The publish workflow
+
+The [publish](.github/workflows/publish.yaml) workflow publishes the snap(s(s) built during the run and after they've been tested.
```
Comment (giusebar) [inline README.md:139]:
```suggestion
The [publish](.github/workflows/publish.yaml) workflow publishes the snap(s) built during the run and after they've been tested.
```

File: .github/workflows/publish.yaml (hunk 19 above)
Comment (artivis) [reply .github/workflows/publish.yaml:43]: Yup debugging :+1:

File: README.md (hunk 20 above)
Comment (artivis) [reply README.md:80]:
```suggestion
as its name suggests,
```

File: .github/workflows/snap.yaml (hunk 18 above)
Comment (artivis) [reply .github/workflows/snap.yaml:99]: It may be undefined on purpose (e.g. no publishing)

File: .github/workflows/snap.yaml (hunk 14 above)
Comment (artivis) [reply .github/workflows/snap.yaml]: Done.

**Review by Guillaumebeuzeboc — APPROVED** [review]

File: .github/workflows/build.yaml (hunk 15 above)
Comment (artivis) [reply .github/workflows/build.yaml:45]: Will change at the very last minute :+1:

**Review by giusebar — APPROVED** [review]

File: .github/workflows/snap.yaml [hunk 23]
```diff
@@ -3,173 +3,118 @@ name: snap
...
+    if: needs.check-snapstore-login.outputs.defined == 'true' && (github.event_name == 'push' || github.event_name == 'workflow_dispatch')
 
   cleanup:
     if: ${{ always() && !cancelled() && needs.build.result == 'success' && inputs.cleanup == true }}
     runs-on: ubuntu-latest
     needs: [build, publish]
     steps:
       - name: Delete artifact
-        uses: geekyeggo/delete-artifact@v4
+        uses: geekyeggo/delete-artifact@v5
```
Comment (MirkoFerrati) [inline .github/workflows/snap.yaml:117]: @Guillaumebeuzeboc how were the rules for self-hosted runners? can we use external github actions? reusable-workflows were forbidden but actions were allowed, right?

Comment (Guillaumebeuzeboc) [reply .github/workflows/snap.yaml:117]: yes.

Comment (Guillaumebeuzeboc) [reply .github/workflows/snap.yaml:117]: I believe GH actions benefit from some security checks that reusable-workflows don't.

**Review by MirkoFerrati — APPROVED** [review]
lgtm!

**Pull request merged**